### DISCUSSION TITLE
Polymorphism and routing in path parameter

### DISCUSSION THREAD

**Discussion by drake-mer** [fastapi/fastapi#8551] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from enum import Enum

import pytest
from fastapi import FastAPI
from starlette.testclient import TestClient

app = FastAPI()


class Meat(str, Enum):
    ham = "ham"
    chicken = "chicken"


class Vegetable(str, Enum):
    carot = "carot"
    cucumber = "cucumber"


@app.get("/flavor/{stuffed_meat}")
def get_meat_sandwich(stuffed_meat: Meat):
    print(f"meat sandwich stuffed with {stuffed_meat}")


@app.get("/flavor/{vege_garniture}")
def get_vege_sandwich(vege_garniture: Vegetable):
    print(f"vegetarian sandwich stuffed with {vege_garniture}")


client = TestClient(app)


@pytest.mark.parametrize("flavor", list(Vegetable) + list(Meat))
def test_sandwich_polymorphism(flavor):
    r = client.get(f"/flavor/{flavor}")
    assert r.status_code == 200

```

### Description

If I create two routes, but I specify explicitely that allowed values for one route is of different type for the two controllers, fastapi doesn't know how to route. It get through the first registered path and thus generate a 422.

### Operating System

Linux

### FastAPI Version

0.70.1

### Python Version

3.10.1

**Comment by dstlny** [comment]
Pretty sure you'd need to write is as:

```python
@app.get("/flavour/{chosen_flavour}")
def get(chosen_flavour: typing.Union[Meat, Vegetable]):
    print(f"sandwich stuffed with {chosen_flavour}")
```

**Comment by rafsaf** [comment]
The solution above seems reasonable, but I would add explanation:

Note, `"/flavor/{stuffed_meat}" == "/flavor/{vege_garniture}"`
both endpoints represent the same paths, look at it from URL perspective, they are both equal to third `"/flavor/{crazy_vegetable}"` or anything you like, in others words, every get request in a manner `/flavor/anystringyouwish` is handled by first controller (and since that second one omitted).

**Comment by drake-mer** [comment]
Thank you for your explanation. Indeed, the solution you propose @dstlny works fine.

However, what was interesting for me was the ability to call two _distinct_ controllers that would be selected according to the detected path parameter type.

Note that for your solution to work, I would have to write something like this:

```python
@app.get("/flavour/{any_flavour}")
def get(any_flavour: typing.Union[Meat, Vegetable]):
    if isinstance(any_flavour, Meat):
        print(f"meat sandwich stuffed with {any_flavour}")
    elif isinstance(any_flavour, Vegetable):
        print(f"vegetarian sandwich stuffed with {any_flavour}")
```

This would work fine, but I hoped to be able not to have to write that kind of code because of better matching on registered controllers.

I think this would be a worthy feature to implement.
(Note: I have updated the example code in the initial issue)

**Comment by mrijken** [comment]
Pyramid has a nice solution to define a path / method multiple times and select the implementation with matching predicates (https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/viewconfig.html).
